File: lua/problem_83.lua
local M = {}

function M.invert_tree(root)
    if not root then return nil end
    root.left, root.right = M.invert_tree(root.right), M.invert_tree(root.left)
    return root
end

local function inorder(node, out)
    if not node then return end
    inorder(node.left, out); out[#out+1] = node.val; inorder(node.right, out)
end

function M.problem83()
    print("Problem 83:")
    local root = {val=4,
        left  = {val=2, left={val=1,left=nil,right=nil}, right={val=3,left=nil,right=nil}},
        right = {val=7, left={val=6,left=nil,right=nil}, right={val=9,left=nil,right=nil}},
    }
    M.invert_tree(root)
    local out = {}; inorder(root, out)
    local exp = {9,7,6,4,3,2,1}
    assert(#out == #exp)
    for i = 1, #exp do assert(out[i] == exp[i]) end
    print("All Tests Passed")
end

return M

File: c/problem_83.c
#include "problem_83.h"
#include <stdio.h>
#include <assert.h>

TreeNode83 *invert_tree(TreeNode83 *root) {
    if(!root) return NULL;
    TreeNode83 *tmp = root->left;
    root->left  = invert_tree(root->right);
    root->right = invert_tree(tmp);
    return root;
}

static void inorder83(TreeNode83 *n, int *out, int *i){
    if(!n) return; inorder83(n->left,out,i); out[(*i)++]=n->val; inorder83(n->right,out,i);
}

void problem83(void) {
    printf("Problem 83:\n");
    TreeNode83 n1={1,NULL,NULL},n3={3,NULL,NULL},n6={6,NULL,NULL},n9={9,NULL,NULL};
    TreeNode83 n2={2,&n1,&n3}, n7={7,&n6,&n9};
    TreeNode83 root={4,&n2,&n7};
    invert_tree(&root);
    int out[8],i=0; inorder83(&root,out,&i);
    int exp[]={9,7,6,4,3,2,1};
    assert(i==7);
    for(int j=0;j<7;j++) assert(out[j]==exp[j]);
    printf("All Tests Passed\n");
}
